### ISSUE TITLE
Fix S2589 FP: Custom Equals(null) is wrongly assumed to always return false

### ISSUE THREAD

**Issue by fiotti** [SonarSource/sonar-dotnet#8382]
### Description

False positive [S2589](https://sonarsource.github.io/rspec/#/rspec/S2589/csharp): when passing `null` to a method `bool Equals(Something? other)` on a `struct`, Sonar makes the wrong assumption that the method will always return `false`.

### Repro steps

```csharp
struct SonarTest {
    static readonly bool s_ignoreMe = true;

    public bool Equals(string? other) => true;

    public bool Test(object? obj) =>
        Equals((string?)null) // S2589: change this condition so that it doesn't always evaluate to 'False'. 
        || s_ignoreMe;
}
```

### Expected behavior

I didn't expect any diagnostics to be emitted.

### Actual behavior

False positive.

### Known workarounds

```csharp
#pragma warning disable S2589
```

### Related information

* SonarLint for Visual Studio 2022, version 7.4.0.80741
* Visual Studio Professional 2022 (64 bit), version 17.7.6
* dotnet 7.0.403
* MSBuild version 17.7.2+d6990bcfa for .NET Framework, version 17.7.2.37605
* Windows 11 Pro 22H2 22621.2715

**Comment by mary-georgiou-sonarsource** [comment]
Hello @fiotti,

This is the `Equals` method, which should always be false if the input is `null`.
The engine does not check the overridden implementation, unfortunately, so from the engine's point of view and its capabilities at this point in time this is a true positive.

However, the fact that it does not raise in the case of a class for example is a false negative and I'll add an issue to our backlog for this.

Please, let me know if this is what you intended in your example (raising an issue with the specific `Equals` implementation.

**Comment by mary-georgiou-sonarsource** [comment]
Hello @fiotti ,

Checking better, we can actually avoid raising on `Equals` in such cases. I'll open an FP issue and add it to our backlog.
Thanks a lot and apologies for the confusion.
